source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -97,100 +122,164 @@ absl::optional<OpCodes> DecoderImpl::decodeOnData(Buffer::Instance& data, uint64
+    status = parseMultiRequest(data, offset, len.value());
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(status,
+                                            fmt::format("parseMultiRequest: {}", status.message()));
     break;
   case OpCodes::Reconfig:
-    parseReconfigRequest(data, offset, len);
+    status = parseReconfigRequest(data, offset, len.value());
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("parseReconfigRequest: {}", status.message()));
     break;
   case OpCodes::SetWatches:
-    parseSetWatchesRequest(data, offset, len);
+    status = parseSetWatchesRequest(data, offset, len.value());
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("parseSetWatchesRequest: {}", status.message()));
     break;
   case OpCodes::SetWatches2:
-    parseSetWatches2Request(data, offset, len);
+    status = parseSetWatches2Request(data, offset, len.value());
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("parseSetWatches2Request: {}", status.message()));
     break;
   case OpCodes::AddWatch:
-    parseAddWatchRequest(data, offset, len);
+    status = parseAddWatchRequest(data, offset, len.value());
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("parseAddWatchRequest: {}", status.message()));
     break;
   case OpCodes::CheckWatches:
-    parseXWatchesRequest(data, offset, len, OpCodes::CheckWatches);
+    status = parseXWatchesRequest(data, offset, len.value(), OpCodes::CheckWatches);
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("parseXWatchesRequest (check watches): {}", status.message()));
     break;
   case OpCodes::RemoveWatches:
-    parseXWatchesRequest(data, offset, len, OpCodes::RemoveWatches);
+    status = parseXWatchesRequest(data, offset, len.value(), OpCodes::RemoveWatches);
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("parseXWatchesRequest (remove watches): {}", status.message()));
     break;
   case OpCodes::GetEphemerals:
-    callbacks_.onGetEphemeralsRequest(pathOnlyRequest(data, offset, len));
+    status = callbacks_.onGetEphemeralsRequest(pathOnlyRequest(data, offset, len.value()));
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("onGetEphemeralsRequest: {}", status.message()));
     break;
   case OpCodes::GetAllChildrenNumber:
-    callbacks_.onGetAllChildrenNumberRequest(pathOnlyRequest(data, offset, len));
+    status = callbacks_.onGetAllChildrenNumberRequest(pathOnlyRequest(data, offset, len.value()));
+    RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(
+        status, fmt::format("onGetAllChildrenNumberRequest: {}", status.message()));
     break;
   case OpCodes::Close:
     callbacks_.onCloseRequest();
     break;
   default:
-    throw EnvoyException(fmt::format("Unknown opcode: {}", enumToSignedInt(opcode)));
+    ENVOY_LOG(debug, "zookeeper_proxy: decodeOnData exception: unknown opcode {}",
```

Comment:
For all the error handling, logging `exception` is misleading, probably `failed`?